File: server/src/routes/candidates.ts
import { FastifyInstance } from "fastify";
import { z } from "zod";
import { supabaseAdmin } from "../database/client.js";

const candidateSchema = z.object({
  name: z.string().min(1),
  email: z.string().email(),
  phone: z.string().optional(),
  title: z.string().optional(),
  skills: z.array(z.string()).default([]),
  experience: z.string().optional(),
  education: z.string().optional(),
  cvText: z.string().optional(),
  githubUrl: z.string().url().optional(),
  portfolioUrl: z.string().url().optional(),
});

export async function candidatesRoutes(app: FastifyInstance) {
  app.get("/", async (request, reply) => {
    const user = (request as any).user;
    const orgId = user.organizationId;
    const { data, error } = await supabaseAdmin.from("candidates").select("*").eq("organization_id", orgId).order("created_at", { ascending: false });
    if (error) return reply.status(500).send({ error: error.message });
    return { candidates: data };
  });

  app.get("/:id", async (request, reply) => {
    const { id } = request.params as { id: string };
    const { data, error } = await supabaseAdmin.from("candidates").select("*").eq("id", id).single();
    if (error) return reply.status(404).send({ error: "Candidate not found" });
    return data;
  });

  app.post("/", async (request, reply) => {
    const user = (request as any).user;
    const parsed = candidateSchema.safeParse(request.body);
    if (!parsed.success) return reply.status(400).send({ error: "Validation failed", details: parsed.error.flatten() });

    const { data, error } = await supabaseAdmin.from("candidates").insert({
      ...parsed.data,
      organization_id: user.organizationId,
    }).select().single();

    if (error) return reply.status(500).send({ error: error.message });
    return reply.status(201).send(data);
  });
}

File: server-python/routes/candidates.py
from fastapi import APIRouter, HTTPException, Request
from auth import get_current_user
from database import db
import json

router = APIRouter(prefix="/api/candidates", tags=["candidates"])


@router.get("")
@router.get("/")
async def list_candidates(request: Request):
    user = await get_current_user(request)
    org_id = user.get("organizationId")
    rows = db.fetchall("SELECT * FROM candidates WHERE organization_id = %s ORDER BY created_at DESC", (org_id,))
    for r in rows:
        if "skills" in r and isinstance(r["skills"], str):
            try:
                r["skills"] = json.loads(r["skills"])
            except Exception:
                r["skills"] = []
    return {"candidates": rows}


@router.get("/{candidate_id}")
async def get_candidate(candidate_id: str):
    row = db.fetchone("SELECT * FROM candidates WHERE id = %s", (candidate_id,))
    if not row:
        raise HTTPException(status_code=404, detail="Candidate not found")
    return row


@router.post("")
@router.post("/")
async def create_candidate(request: Request):
    user = await get_current_user(request)
    body = await request.json()
    row = db.insert("candidates", {
        "name": body.get("name", ""),
        "email": body.get("email", ""),
        "phone": body.get("phone"),
        "skills": json.dumps(body.get("skills", [])),
        "experience": body.get("experience"),
        "github_url": body.get("githubUrl"),
        "organization_id": user.get("organizationId"),
    })
    return row
